Fix disk space check when selecting model

fbricon opened the pull request:
Use [check-disk-space ](https://www.npmjs.com/package/check-disk-space) library to check disk space on `~`. 
Also fixed checking required space when model is already present.

Fixes https://github.com/Granite-Code/granite-code/issues/58

halfline commented on extensions/vscode/src/granite/utils/sysUtils.ts:
```diff
@@ -16,20 +17,12 @@ export function getSystemCPUs() {
-      };
-    }
-    console.error("Main volume not found");
+    const ds = await checkDiskSpace(os.homedir());
```
is os.homedir() right for ollama on windows ? I guess it probably is, but just checking. This function looks much nicer!

fbricon replied:
yes it should work

halfline commented on gui/src/granite/GraniteWizard.tsx:
```diff
@@ -1,38 +1,33 @@
 import { isHighEndMachine, SystemInfo } from "core/granite/commons/sysInfo";
+import { formatSize } from "core/granite/commons/textUtils";
 import {
-  OLLAMA_STEP,
-  MODELS_STEP,
   FINAL_STEP,
+  MODELS_STEP,
+  OLLAMA_STEP,
   WizardState,
 } from "core/granite/commons/wizardState";
-import { vscode } from "./utils/vscode";
-import { ProgressData } from "core/granite/commons/progressData";
-import { formatSize } from "core/granite/commons/textUtils";
-import { DEFAULT_MODEL_INFO } from "core/granite/commons/modelInfo";
-import {
-  DEFAULT_MODEL_GRANITE_LARGE,
-  DEFAULT_MODEL_GRANITE_SMALL,
-} from "core/config/default";
-import { LocalModelSize } from "core";
-import { ProgressBlock } from "./ProgressBlock";
+import React, {
+  createContext,
+  ReactNode,
+  useContext,
+  useEffect,
+  useRef,
+  useState,
+} from "react";
+import { VSCodeButton } from "../components/VSCodeButton";
 import { DiagnosticMessage } from "./DiagnosticMessage";
+import { ProgressBlock } from "./ProgressBlock";
+import { StatusCheck } from "./StatusCheck";
+import { vscode } from "./utils/vscode";
```
This part of the diff is hard to read because you're moving a bunch of stuff around and I guess maybe dropping unnecessary code? It doesn't seem all the relevant to the change at hand, so probably shouldn't be in the same commit. Can you split it out?

fbricon replied:
done

halfline commented on gui/src/granite/GraniteWizard.tsx:
```diff
@@ -428,8 +428,8 @@ const ModelSelectionStep: React.FC<StepProps> = (props) => {
     if (systemInfo && systemInfo.diskSpace) {
-      const { freeDiskSpace } = systemInfo.diskSpace;
-      const requiredDiskSpace = getRequiredSpace(selectedModel);
+      let { freeDiskSpace } = systemInfo.diskSpace;
```
why did you change this to `let`?

halfline commented on gui/src/granite/GraniteWizard.tsx:
```diff
@@ -115,7 +112,9 @@ export const WizardProvider: React.FC<WizardProviderProps> = ({ children }) => {
-  const [modelStatus, setModelStatus] = useState<StatusValue>("missing");
+  const [modelStatuses, setModelStatuses] = useState<Map<string, ModelStatus>>(
+    new Map(),
+  );
```
fwiw, the word statuses looks clunky to me. I realize it's a valid word and there's prior art in the code, but... `setModelStatuses` to me could mean there is more than one status for one model, while `setModelStatus` could mean one status for one model or one status for each model. So I actually like the old name better. another idea is `statusByModel` and `setStatusByModel`

anyway, it's fine as is, it just doesn't quite read smoothly to me, so I thought I'd point it out in case you agree.

fbricon replied:
renamed to statusByModel / setStatusByModel

halfline commented on gui/src/granite/StatusCheck.tsx:
```diff
@@ -1,7 +1,6 @@
 import {
   VscArrowCircleDown,
   VscCircleLarge,
-  VscCircleLargeFilled,
```
unrelated change

halfline commented on gui/src/granite/GraniteWizard.tsx:
```diff
@@ -824,15 +826,20 @@ const WizardContent: React.FC = () => {
-    return sum + (modelInfo ? modelInfo.size : 0);
-  }, 0);
+  const reqSpace = models
+    .filter((model) => modelStatuses.get(model) === ModelStatus.missing)
```
so this part should probably be a separate commit too. maybe the pr could be 3 commits:

1. one that cleans up and rearranges imports
2. one that switches over to looking at home directory instead of system volume
3. one that removes already downloaded models from the required space computation.

fbricon replied:
done

halfline reviewed (approved):
looks great.

fbricon committed:
gui: Fix code formatting with prettier

There are some minor style problems in the
setup wizard related code.

- Some of the imports are in the wrong order
- Not all lines wrap properly

This commit addresses those problems by running
the code through Prettier and letting it figure
it out.

Signed-off-by: Fred Bricon <fbricon@gmail.com>

fbricon committed:
gui: Fix disk space calculation on some platforms

The setup wizard needs to know how much disk is available
on the user's system so it can inform the user if there
is insufficient disk space to download the (rather large)
Granite models.

The code currently does this using the systeminformation
module to check the available space on the root filesystem.

That isn't exactly correct, though, because models are
stored in the users directory, which may be on a different
filesystem altogether.

Furthermore, systeminformation doesn't handle btrfs systems
properly, since it keys output off the volume name, and the
volume name is shared between subvolumes with btrfs.

This commit addresses both problems by switching the code
over to use the check-disk-space module, which is more
form fit for our purposes anyway.

Signed-off-by: Fred Bricon <fbricon@gmail.com>

fbricon committed:
gui: Deduct existing models from diskspace needs

The setup wizard reports to the user if there is insufficent
diskspace to download the Granite model the user has selected.

If there is insufficient diskspace to proceed it stops the
user from continuing.

Unfortunately, there is a oversight in the code where it fails
to consider that the user may already have a model downloaded.

In that case, the model doesn't need new diskspace.

This commit fixes that problem by reporting the download status
of each model to the wizard and then ignoring already downloaded
models when calculating diskspace needs.

Signed-off-by: Fred Bricon <fbricon@gmail.com>

halfline commented:
looks great. the only thing i would do is better commit messages:

for example:

```commit 626a353fa77f5c69534a1d365b8432ab5973e15d (HEAD -> fix-diskspace-check)
Author: Fred Bricon <fbricon@gmail.com>
Date:   Tue Mar 11 11:18:24 2025 +0100

    gui: Deduct existing models from diskspace needs
    
    The setup wizard reports to the user if there is insufficent
    diskspace to download the Granite model the user has selected.
    
    If there is insufficient diskspace to proceed it stops the
    user from continuing.
    
    Unfortunately, there is a oversight in the code where it fails
    to consider that the user may already have a model downloaded.
    
    In that case, the model doesn't need new diskspace.
    
    This commit fixes that problem by reporting the download status
    of each model to the wizard and then ignoring already downloaded
    models when calculating diskspace needs.
    
    Signed-off-by: Fred Bricon <fbricon@gmail.com>

commit 922d0f84568401dfd2603bb2fc585e946091fbb9
Author: Fred Bricon <fbricon@gmail.com>
Date:   Tue Mar 11 10:52:26 2025 +0100

    gui: Fix disk space calculation on some platforms
    
    The setup wizard needs to know how much disk is available
    on the user's system so it can inform the user if there
    is insufficient disk space to download the (rather large)
    Granite models.
    
    The code currently does this using the systeminformation
    module to check the available space on the root filesystem.
    
    That isn't exactly correct, though, because models are
    stored in the users directory, which may be on a different
    filesystem altogether.
    
    Furthermore, systeminformation doesn't handle btrfs systems
    properly, since it keys output off the volume name, and the
    volume name is shared between subvolumes with btrfs.
    
    This commit addresses both problems by switching the code
    over to use the check-disk-space module, which is more
    form fit for our purposes anyway.
    
    Signed-off-by: Fred Bricon <fbricon@gmail.com>

commit 4d9a4b484e309b630e1ac081b7e5d0243f6b99c6
Author: Fred Bricon <fbricon@gmail.com>
Date:   Tue Mar 11 10:20:18 2025 +0100

    gui: Fix code formatting with prettier
    
    There are some minor style problems in the
    setup wizard related code.
    
    - Some of the imports are in the wrong order
    - Not all lines wrap properly
    
    This commit addresses those problems by running
    the code through Prettier and letting it figure
    it out.
    
    Signed-off-by: Fred Bricon <fbricon@gmail.com>
```

but feel free to merge whenever
